From 84d2d608aa46588034d2012d43ee283f66b0d8b8 Mon Sep 17 00:00:00 2001
From: anonym <anonym@riseup.net>
Date: Wed, 7 Oct 2020 15:50:23 +0200
Subject: [PATCH] Add pref for whether we accept OAuth2 during
 autoconfiguration.

For many providers JavaScript is required for OAuth2 to work; with it
disabled autoconfiguration then result in a terrible UX (e.g. the web
login fails, has to manually alter the authentication method). Let's
provide a pref that discards OAuth2 configurations so e.g. extensions
that disables JavaScript (like TorBirdy) can provide a workaround.
---
 .../accountcreation/content/emailWizard.js    | 72 ++++++++++---------
 .../accountcreation/content/readFromXML.js    | 13 ++++
 comm/mailnews/mailnews.js                     |  2 +
 3 files changed, 53 insertions(+), 34 deletions(-)

diff --git a/comm/mail/components/accountcreation/content/emailWizard.js b/comm/mail/components/accountcreation/content/emailWizard.js
index 18c2b514894..487f9aefae0 100644
--- a/comm/mail/components/accountcreation/content/emailWizard.js
+++ b/comm/mail/components/accountcreation/content/emailWizard.js
@@ -1467,23 +1467,25 @@ EmailConfigWizard.prototype = {
     }
     this.fillPortDropdown(config.incoming.type);
 
-    // If the incoming server hostname supports OAuth2, enable OAuth2 for it.
-    let iDetails = OAuth2Providers.getHostnameDetails(config.incoming.hostname);
-    e("in-authMethod-oauth2").hidden = !iDetails;
-    if (iDetails) {
-      gEmailWizardLogger.info(
-        "OAuth2 details for incoming server " +
-          config.incoming.hostname +
-          " is " +
-          iDetails
-      );
-      config.incoming.oauthSettings = {};
-      [
-        config.incoming.oauthSettings.issuer,
-        config.incoming.oauthSettings.scope,
-      ] = iDetails;
-      this._currentConfig.incoming.oauthSettings =
-        config.incoming.oauthSettings;
+    if (Services.prefs.getBoolPref("mailnews.auto_config.account_constraints.allow_oauth2")) {
+      // If the incoming server hostname supports OAuth2, enable OAuth2 for it.
+      let iDetails = OAuth2Providers.getHostnameDetails(config.incoming.hostname);
+      e("in-authMethod-oauth2").hidden = !iDetails;
+      if (iDetails) {
+        gEmailWizardLogger.info(
+          "OAuth2 details for incoming server " +
+            config.incoming.hostname +
+            " is " +
+            iDetails
+        );
+        config.incoming.oauthSettings = {};
+        [
+          config.incoming.oauthSettings.issuer,
+          config.incoming.oauthSettings.scope,
+        ] = iDetails;
+        this._currentConfig.incoming.oauthSettings =
+          config.incoming.oauthSettings;
+      }
     }
 
     // outgoing server
@@ -1508,23 +1510,25 @@ EmailConfigWizard.prototype = {
       this.adjustOutgoingPortToSSLAndProtocol(config);
     }
 
-    // If the smtp hostname supports OAuth2, enable OAuth2 for it.
-    let oDetails = OAuth2Providers.getHostnameDetails(config.outgoing.hostname);
-    e("out-authMethod-oauth2").hidden = !oDetails;
-    if (oDetails) {
-      gEmailWizardLogger.info(
-        "OAuth2 details for outgoing server " +
-          config.outgoing.hostname +
-          " is " +
-          oDetails
-      );
-      config.outgoing.oauthSettings = {};
-      [
-        config.outgoing.oauthSettings.issuer,
-        config.outgoing.oauthSettings.scope,
-      ] = oDetails;
-      this._currentConfig.outgoing.oauthSettings =
-        config.outgoing.oauthSettings;
+    if (Services.prefs.getBoolPref("mailnews.auto_config.account_constraints.allow_oauth2")) {
+      // If the smtp hostname supports OAuth2, enable OAuth2 for it.
+      let oDetails = OAuth2Providers.getHostnameDetails(config.outgoing.hostname);
+      e("out-authMethod-oauth2").hidden = !oDetails;
+      if (oDetails) {
+        gEmailWizardLogger.info(
+          "OAuth2 details for outgoing server " +
+            config.outgoing.hostname +
+            " is " +
+            oDetails
+        );
+        config.outgoing.oauthSettings = {};
+        [
+          config.outgoing.oauthSettings.issuer,
+          config.outgoing.oauthSettings.scope,
+        ] = oDetails;
+        this._currentConfig.outgoing.oauthSettings =
+          config.outgoing.oauthSettings;
+      }
     }
 
     // populate fields even if existingServerKey, in case user changes back
diff --git a/comm/mail/components/accountcreation/content/readFromXML.js b/comm/mail/components/accountcreation/content/readFromXML.js
index d0fa70846c9..3c3991ddb7a 100644
--- a/comm/mail/components/accountcreation/content/readFromXML.js
+++ b/comm/mail/components/accountcreation/content/readFromXML.js
@@ -33,6 +33,8 @@ function readFromXML(clientConfigXML, subSource) {
   function array_or_undef(value) {
     return value === undefined ? [] : value;
   }
+  var allow_oauth2 =
+    Services.prefs.getBoolPref("mailnews.auto_config.account_constraints.allow_oauth2");
   var exception;
   if (
     typeof clientConfigXML != "object" ||
@@ -128,6 +130,12 @@ function readFromXML(clientConfigXML, subSource) {
             NTLM: Ci.nsMsgAuthMethod.NTLM,
             OAuth2: Ci.nsMsgAuthMethod.OAuth2,
           });
+
+          if (!allow_oauth2 && iO.auth == Ci.nsMsgAuthMethod.OAuth2) {
+            iO.auth = null;
+            continue;
+          }
+
           break; // take first that we support
         } catch (e) {
           exception = e;
@@ -271,6 +279,11 @@ function readFromXML(clientConfigXML, subSource) {
             OAuth2: Ci.nsMsgAuthMethod.OAuth2,
           });
 
+          if (!allow_oauth2 && oO.auth == Ci.nsMsgAuthMethod.OAuth2) {
+            oO.auth = null;
+            continue;
+          }
+
           break; // take first that we support
         } catch (e) {
           exception = e;
diff --git a/comm/mailnews/mailnews.js b/comm/mailnews/mailnews.js
index ddebe393a33..ff2ddf5284e 100644
--- a/comm/mailnews/mailnews.js
+++ b/comm/mailnews/mailnews.js
@@ -977,6 +977,8 @@ pref("mailnews.auto_config.guess.enabled", true);
 pref("mailnews.auto_config.guess.sslOnly", false);
 // The timeout (in seconds) for each guess
 pref("mailnews.auto_config.guess.timeout", 10);
+// Whether we allow fetched configurations using OAuth2.
+pref("mailnews.auto_config.account_constraints.allow_oauth2", true);
 // Work around bug 1454325 by disabling mimetype mungling in XmlHttpRequest
 pref("dom.xhr.standard_content_type_normalization", false);
 
-- 
2.31.0

